ddtrace/tracer/sqlcomment.go

```diff
@@ -188,7 +189,104 @@ func commentQuery(query string, tags map[string]string) string {
 	return b.String()
 }
 
-// Extract is not implemented on SQLCommentCarrier
+// Extract parses for key value attributes in a sql query injected with trace information in order to build a span context
 func (c *SQLCommentCarrier) Extract() (ddtrace.SpanContext, error) {
-	return nil, nil
+	var ctx *spanContext
+	// There may be multiple comments within the sql query, so we must identify which one contains trace information.
+	// We look at each comment until we find one that contains a traceparent
+	if traceComment, found := findTraceComment(c.Query); found {
+		var err error
+		if ctx, err = spanContextFromTraceComment(traceComment); err != nil {
+			return nil, err
+		}
+	} else {
+		return nil, ErrSpanContextNotFound
+	}
+	if ctx.traceID.Empty() || ctx.spanID == 0 {
+		return nil, ErrSpanContextNotFound
+	}
+	return ctx, nil
+}
+
+// spanContextFromTraceComment looks for specific kv pairs in a comment containing trace information.
+// It returns a span context with the appropriate attributes
+func spanContextFromTraceComment(c string) (*spanContext, error) {
+	var ctx spanContext
+	kvs := strings.Split(c, ",")
+	for _, unparsedKV := range kvs {
+		if splitKV := strings.Split(unparsedKV, "="); len(splitKV) == 2 {
+			key := splitKV[0]
+			value := strings.Trim(splitKV[1], "'")
+			switch key {
+			case sqlCommentTraceParent:
+				traceID, spanID, sampled, err := decodeTraceParent(value)
+				if err != nil {
+					return nil, err
+				}
+				ctx.traceID.SetLower(traceID)
+				ctx.spanID = spanID
+				ctx.setSamplingPriority(sampled, samplernames.Unknown)
+			default:
+			}
+		} else {
+			return nil, ErrSpanContextCorrupted
+		}
+	}
+	return &ctx, nil
+}
+
+// decodeTraceParent decodes trace parent as per the w3c trace context spec (https://www.w3.org/TR/trace-context/#version).
+func decodeTraceParent(traceParent string) (traceID uint64, spanID uint64, sampled int, err error) {
+	version := traceParent[0:2]
```

Comment:
Does there need to be a check to ensure that the `traceparent` is long enough here?

Reply:
I like that idea too. The current version (that future versions should be backward compatible with) is 55 bytes long (unless I miscounted) so we could check that `len(traceParent) >= 55`.